### PR TITLE
Update to bionic

### PR DESCRIPTION
**Pull request by fasmat**
Updated linus-kitchen to Bionic

### CODE REVIEW

File: cookbooks/vm/recipes/git.rb [hunk 1]
```diff
@@ -20,8 +20,3 @@
   mode '0644'
   action :create_if_missing
 end
-
-bashd_entry 'setup-git-ps1-prompt' do
```
[the remark this answers is not in the document]
Comment (fasmat) [reply cookbooks/vm/recipes/git.rb]: Unfortunately not, the bashrc_manager resource doesn't offer to parse and store template files. One could parse the template manually into a string and pass that to the bashrc_manager resource.
My ruby skills are not good enough to implement that myself, maybe you can help me with that?

[the remark this answers is not in the document]
Comment (fasmat) [reply cookbooks/vm/recipes/git.rb]: Found a workaround :-)

[the remark this answers is not in the document]
Comment (tknerr) [reply cookbooks/vm/recipes/git.rb]: heh, nice! :)

[the remark this answers is not in the document]
Comment (tknerr) [reply cookbooks/vm/recipes/git.rb]: On second look: I have no clue why I added the `git_ps1.erb` as a template, because there are no interpolated variables in it at all. Oh yes I know, it was because the former [bashd cookbook](https://github.com/jakshi/bashd) was an .erb file...

Since it is not templated anyway, you could as well
a) define the whole content inline (I would probably use a [squiggly heredoc](http://www.virtuouscode.com/2016/01/06/about-the-ruby-squiggly-heredoc-syntax/))
b) simply use `IO.read(filename)` instead of the newly added `from_template` function

[the remark this answers is not in the document]
Comment (fasmat) [reply cookbooks/vm/recipes/git.rb]: I went for b) and updated the pull request

**Comment by fasmat** [comment]
Now with green build :-)

**Comment by tknerr** [comment]
@fasmat this PR is pure gold, thanks so much for the continued dedication! Green build, yeeeesssssss!

I owe you a 🍻 or two now! :)

**Comment by tknerr** [comment]
I have a few comments / questions, but basically happy to merge this as-is now 👍

File: Vagrantfile [hunk 2]
```diff
@@ -2,27 +2,33 @@
 Vagrant::configure("2") do |config|
 
   # configure the basebox
-  config.vm.box = "tknerr/ubuntu1604-desktop"
-  config.vm.box_version = "2.0.27.1"
+  config.vm.box = "fasmat/ubuntu1804-desktop"
```
Comment (tknerr) [inline Vagrantfile]: Sweet, that's how you end up with your own basebox! :)

I found https://app.vagrantup.com/fasmat/boxes/ubuntu1804-desktop here, but was wondering how you built that basebox in the first place. In case it was build using packer, a reference to the repo with the packer template that was used to create the box would be helpful (in the basebox description on app.vagrantup.com I mean).

My https://app.vagrantup.com/tknerr/boxes/ubuntu1604-desktop for example was based on https://github.com/boxcutter/ubuntu.

Happy to use yours, was just wondering how to find out how it was built and what's included

File: Vagrantfile [hunk 3]
```diff
@@ -2,27 +2,33 @@
...
 
   # virtualbox specific customizations
   config.vm.provider :virtualbox do |vbox, override|
-    vbox.customize ["modifyvm", :id,
-      "--name", "Linus Kitchen",
-      "--memory", 4096,
-      "--cpus", Etc.nprocessors
-    ]
+    vbox.customize ["modifyvm", :id, "--name", "Linus Kitchen"]
+    vbox.customize ["modifyvm", :id, "--memory", 8192]
```
Comment (tknerr) [inline Vagrantfile]: I was tempted to increase to 8GB earlier, but then switched back to 4GB as I was mostly using the docker provider inside. But to honestly run a Windows Virtualbox VM inside, you definitely need the 8GB ;) 👍

File: Vagrantfile [hunk 4]
```diff
@@ -2,27 +2,33 @@
...
+    vbox.customize ["modifyvm", :id, "--cpus", Etc.nprocessors]
+    vbox.customize ["modifyvm", :id, "--vram", 256]
+    vbox.customize ["modifyvm", :id, "--accelerate3d", "off"]
+    vbox.customize ["modifyvm", :id, "--ioapic", "on"]
+    vbox.customize ["modifyvm", :id, "--clipboard", "bidirectional"]
```
Comment (tknerr) [inline Vagrantfile]: Thanks for that, this is super useful! :)

File: cookbooks/vm/recipes/atom.rb [hunk 5]
```diff
@@ -1,12 +1,12 @@
 
-atom_version = '1.17.1'
+atom_version = '1.32.2'
```
Comment (tknerr) [inline cookbooks/vm/recipes/atom.rb]: nowadays I would actually install VS Code, but that would be out of scope for this PR anyway..

File: cookbooks/vm/recipes/git.rb [hunk 6]
```diff
@@ -21,7 +21,7 @@
   action :create_if_missing
 end
 
-bashd_entry 'setup-git-ps1-prompt' do
-  user vm_user
-  source 'git_ps1.erb'
+bashrc_manager 'setup-git-ps1-prompt' do
+  user vm_user	
+  content IO.read("#{run_context.cookbook_collection[cookbook_name].root_dir}/files/default/git_ps1")
```
Comment (tknerr) [inline cookbooks/vm/recipes/git.rb]: thanks for re-adding this! :)

File: cookbooks/vm/spec/integration/recipes/git_spec.rb [hunk 7]
```diff
@@ -34,7 +38,7 @@
         st: 'status',
         unstage: 'reset HEAD --',
         slog: 'log --pretty=oneline --abbrev-commit',
-        graph: 'log --all --oneline --graph --decorate'
+        graph: 'log --all --oneline --graph --decorate',
```
Comment (tknerr) [inline cookbooks/vm/spec/integration/recipes/git_spec.rb]: remove trailing comma?

File: scripts/setup-vm-user.sh [hunk 8]
```diff
@@ -6,18 +6,21 @@ LOGIN_PASS=$2
...
-  echo "$LOGIN_USER ALL=(ALL) NOPASSWD: ALL" | sudo tee /etc/sudoers.d/$LOGIN_USER
+# create user, set password and add it to the usual groups
+adduser $LOGIN_USER --gecos "" --home "/home/$LOGIN_USER" --disabled-password
+echo "$LOGIN_USER:$LOGIN_PASS" | chpasswd
+usermod -a -G adm,cdrom,sudo,dip,plugdev,lpadmin,sambashare $LOGIN_USER
 
-  # set the new user as the default in the login screen and lock the current (vagrant user's) screen session
-  if [[ $(which dm-tool) ]]; then
-    echo -e "[SeatDefaults]\nautologin-user=$LOGIN_USER" | sudo tee /etc/lightdm/lightdm.conf
-    XDG_SEAT_PATH=/org/freedesktop/DisplayManager/Seat0 dm-tool lock
```
Comment (tknerr) [inline scripts/setup-vm-user.sh]: I remember that I locked the screen on purpose, but I don't remember why... hmmmm (*slaps himself for not writing a proper comment*)

File: scripts/update-vm.sh [hunk 9]
```diff
@@ -24,11 +24,7 @@ setup_chefdk() {
     echo "ChefDK $CHEFDK_VERSION already installed"
   else
     step "Downloading and installing ChefDK $CHEFDK_VERSION"
-    mkdir -p $DOWNLOAD_DIR
-    local CHEFDK_DEB=chefdk_$CHEFDK_VERSION-1_amd64.deb
-    local CHEFDK_URL=https://packages.chef.io/files/stable/chefdk/$CHEFDK_VERSION/ubuntu/16.04/$CHEFDK_DEB
-    [[ -f $DOWNLOAD_DIR/$CHEFDK_DEB ]] || sudo wget --no-verbose -O $DOWNLOAD_DIR/$CHEFDK_DEB $CHEFDK_URL
-    sudo dpkg -i $DOWNLOAD_DIR/$CHEFDK_DEB
+    curl -L https://omnitruck.chef.io/install.sh | sudo bash -s -- -P chefdk -c stable -v $CHEFDK_VERSION -d /tmp/vagrant-cache
```
Comment (tknerr) [inline scripts/update-vm.sh]: FYI - the reason I had the more complicated version earlier was that I wanted to have this script work even if I was completely offline-

I'm fine with the single line curl / bash pipe too (it's the more idiomatic and bulletproof way to install chef anyway), so let's keep it!

File: cookbooks/vm/recipes/vagrant.rb [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 
 vagrant_plugins = {
   'vagrant-cachier' => '1.2.1',
-  'vagrant-berkshelf' => '5.1.1',
+  'vagrant-berkshelf' => '5.1.2',
   'vagrant-omnibus' => '1.5.0',
   'vagrant-toplevel-cookbooks' => '0.2.4',
   'vagrant-managed-servers' => '0.8.0',
-  'vagrant-lxc' => '1.2.3'
+  'vagrant-lxc' => '1.4.3',
```
Comment (tknerr) [inline cookbooks/vm/recipes/vagrant.rb]: remove trailing comma?

File: scripts/setup-vm-user.sh (hunk 8 above)
Comment (tknerr) [reply scripts/setup-vm-user.sh]: Based on the commit message from 2d5f3d143264047bf19f22b3118bc7fdb989cc00, I wanted to avoid that you accidentally use the VM with the vagrant user, where you actually should log out and log in as `$LOGIN_USER` instead. Locking the screen with the auto login user `$LOGIN_USER` somehow solved that.

Comment (tknerr) [reply scripts/setup-vm-user.sh]: If the screen locking still works that way, I'd be happy to keep it

**Comment by tknerr** [comment]
@fasmat to summarize the review comments from above:

1. awesome work, thanks!!
2. found 2 trailing commas to be removed
3. if screen locking via dm-tool (or gdm3 equivalent) would still work that would be nice 

Let me know if you have some more energy left for this or if I should jump in for that. 

Assuming that you consider it complete otherwise, I'd be happy to get this merged

File: scripts/setup-vm-user.sh [hunk 11]
```diff
@@ -6,18 +6,21 @@ LOGIN_PASS=$2
...
-  fi
-fi
+# ensure the new user can do passwordless sudo
+echo "$LOGIN_USER ALL=(ALL) NOPASSWD: ALL" | sudo tee /etc/sudoers.d/$LOGIN_USER
+
+# set the new user as the default in the login screen
+mkdir -p /etc/gdm3
+> /etc/gdm3/custom.conf
```
Comment (tknerr) [inline scripts/setup-vm-user.sh]: Is this a shorter form of `touch`? I noticed this work in bash, but might confuse zsh users ;)

File: Vagrantfile (hunk 2 above)
Comment (fasmat) [reply Vagrantfile]: I forked boxcutter/ubuntu and updated it to be able to create a 18.04 image. I thought I had the link in the box description, but thanks for the hint! I'll add it tomorrow.

See the following pull request for the changes I made to be able to create the base box:
https://github.com/boxcutter/ubuntu/pull/162

File: Vagrantfile (hunk 4 above)
Comment (fasmat) [reply Vagrantfile]: No problem. I played around with these settings for a while and these gave me the best performance. Especially turning off 3d acceleration made the Gnome desktop feel much faster.

File: cookbooks/vm/recipes/atom.rb (hunk 5 above)
Comment (fasmat) [reply cookbooks/vm/recipes/atom.rb]: I'm working on that, just not ready for a pull request yet: https://github.com/fasmat/linus-kitchen/commits/customize-installed-applications

Will come with recipes and roles, so you can customize which applications you want with which plugins.

File: cookbooks/vm/spec/integration/recipes/git_spec.rb (hunk 7 above)
Comment (fasmat) [reply cookbooks/vm/spec/integration/recipes/git_spec.rb]: :-D bad habit from golang. There it is mandatory: https://dave.cheney.net/2014/10/04/that-trailing-comma

I'll remove it again.

File: scripts/setup-vm-user.sh (hunk 11 above)
Comment (fasmat) [reply scripts/setup-vm-user.sh]: not quite, it clears the file if it exists:

```
fasmat@riddler:~$ echo "test" > test.txt
fasmat@riddler:~$ cat test.txt 
test
fasmat@riddler:~$ > test.txt
fasmat@riddler:~$ cat test.txt 
fasmat@riddler:~$ 
```

Comment (fasmat) [reply scripts/setup-vm-user.sh]: The goal is to have the script run idem-potently. Touch would cause the following lines to be appended to the file with every subsequent run.

**Comment by fasmat** [comment]
All review findings fixed and build is back to green :-)

File: Vagrantfile (hunk 2 above)
Comment (tknerr) [reply Vagrantfile]: Sweet, thanks for the info! 👍

File: scripts/setup-vm-user.sh (hunk 11 above)
Comment (tknerr) [reply scripts/setup-vm-user.sh]: Good point, +1 for idempotence. Thanks for elaborating! :)

File: scripts/setup-vm-user.sh [hunk 12]
```diff
@@ -17,10 +17,12 @@ usermod -a -G adm,cdrom,sudo,dip,plugdev,lpadmin,sambashare $LOGIN_USER
...
-# set the new user as the default in the login screen
+# set the new user as the default in the login screen and end the current (vagrant) user's gnome session
 mkdir -p /etc/gdm3
 > /etc/gdm3/custom.conf
 echo "[daemon]" >> /etc/gdm3/custom.conf
 echo "# Enabling automatic login" >> /etc/gdm3/custom.conf
 echo "AutomaticLoginEnable = true" >> /etc/gdm3/custom.conf
 echo "AutomaticLogin = $LOGIN_USER" >> /etc/gdm3/custom.conf
+
+systemctl restart display-manager
```
Comment (tknerr) [inline scripts/setup-vm-user.sh]: so this did the trick to logout the user?

File: cookbooks/vm/recipes/atom.rb (hunk 5 above)
Comment (tknerr) [reply cookbooks/vm/recipes/atom.rb]: Thanks for the heads up! :)

I have only skimmed the changes in the branch, but my first impression was on whether we actually need the configurability / composability there. Personally, I would probably tend towards the opinionated approach and use a fixed runlist with the single `vm::default` recipe. 
Once you start making this configurable, you may run into many more edge cases and discover you have dependencies between the roles, then ordering comes into play, and finally you need the tests to run conditionally depending on whether that role was included or not.
If there are some tools you consider optional, I'd rather have a flag in the json file for this and then conditionally include / exclude that in the default recipe (also saves a lot of role json files).

Just my 5 cents though, and not related to this PR anyway ;)

The VS Code changes Iook promising as well! :)

File: scripts/setup-vm-user.sh (hunk 8 above)
Comment (tknerr) [reply scripts/setup-vm-user.sh]: this was done via `systemctl restart display-manager` now, right?

File: scripts/setup-vm-user.sh (hunk 12 above)
Comment (fasmat) [reply scripts/setup-vm-user.sh]: It restarts the display server, so it will log out the current user (vagrant) and (because of the script above) immediately log in the newly created user.

**Comment by tknerr** [comment]
Thanks @fasmat again! 

I consider this ready to merge (unless there is something you would want to add before -- in that case let me know)

File: scripts/setup-vm-user.sh (hunk 8 above)
Comment (fasmat) [reply scripts/setup-vm-user.sh]: yes, gnome-session-quit only works if called from the gnome session, if you try to call it in an ssh session you'll get an error, furthermore the current approach is independent from the used display manager and should work with any ubuntu version 15.04 or newer

File: cookbooks/vm/recipes/atom.rb (hunk 5 above)
Comment (fasmat) [reply cookbooks/vm/recipes/atom.rb]: For me it's personal preference. I'd like to have different environments for the development work I do currently. Therefore I'd like to have one environment for golang development, one for .NETCore, one for LaTeX, aso. With only the minimal tools included that I need in that environment.

For testing I yet have to figure out how to only run specific tests based on roles. In another chef project I'm using tags (based on the name of provided run-list). If you know of a better way that would be cool :-)

**Comment by fasmat** [comment]
Nope, looks good to me. Go for the merge :-)

File: scripts/setup-vm-user.sh (hunk 12 above)
Comment (tknerr) [reply scripts/setup-vm-user.sh]: Thanks!

File: cookbooks/vm/recipes/atom.rb (hunk 5 above)
Comment (tknerr) [reply cookbooks/vm/recipes/atom.rb]: Tags sound like a good approach. Not done that in Chef yet, but something similar in Ansible once.

Ah, now I see what you are up to. The use case to mix and match the tools for the specific purpose / project you need them is of course perfectly valid. Haven't considered it in that context, because linus-kitchen is meant to be a developer VM specifically for "Infrastructure-as-Code" development, thus it is biased / opinionated by nature.

Still, I pretty much like the idea of a general purpose developer VM "framework" where you can then put together and mix and match the toolchains you need. What I believe we need for this would be a central repository for collecting these "toolchain roles" (along with the tests for them), which we could then simply include or reference in the "developer VM" repos. This would also be a huge leap towards better reusability of such roles. I have some ideas on this, would be happy to discuss and exchange if you want.

Comment (fasmat) [reply cookbooks/vm/recipes/atom.rb]: Sure, you know where to find me ;-)
